### DISCUSSION TITLE
兑换失败，返回1028

### DISCUSSION THREAD

**Discussion by hhhggg233** [Ljzd-PRO/Mys_Goods_Tool#135] [Q&A]
这是1.4.5版本
2023-07-05 19:00:00  INFO  开始发送商品 2023062912693 的兑换请求...
2023-07-05 19:00:00  INFO  开始发送商品 2023062912693 的兑换请求...
2023-07-05 19:00:00  INFO  开始发送商品 2023062912210 的兑换请求...
2023-07-05 19:00:00  INFO  开始发送商品 2023062912693 的兑换请求...
2023-07-05 19:00:00  INFO  开始发送商品 2023062912210 的兑换请求...
2023-07-05 19:00:00  INFO  开始发送商品 2023062912210 的兑换请求...
2023-07-05 19:00:00  INFO  兑换商品：2023062912693 返回结果：
Too Many Requests

2023-07-05 19:00:00  INFO  兑换商品：2023062912210 返回结果：
Too Many Requests

2023-07-05 19:00:00  INFO  兑换商品：2023062912693 返回结果：
{"data":null,"message":"","retcode":1028}

2023-07-05 19:00:00  INFO  兑换商品：2023062912210 返回结果：
Too Many Requests

2023-07-05 19:00:00  INFO  兑换商品：2023062912693 返回结果：
Too Many Requests

2023-07-05 19:00:00  INFO  兑换商品：2023062912210 返回结果：
Too Many Requests

2023-07-05 19:00:00  INFO  商品 2023062912693 兑换失败，可以自行确认。
2023-07-05 19:00:00  INFO  程序运行结束

这是2.0.5版本
07-05 19:00:00 [INFO] mys_goods_tool.api | 米游币商品兑换 - 执行兑换: 用户 77347599 商品 2023062912693 兑换失败，可以自行确认。
07-05 19:00:00 [DEBUG] mys_goods_tool.api | 网络请求返回: {"data":null,"message":"","retcode":1028}
07-05 19:00:00 [ERROR] mys_goods_tool.exchange_mode | 用户 77347599 - 【原神】原石*150 - 线程 1 - 兑换失败
07-05 19:00:00 [INFO] mys_goods_tool.api | 米游币商品兑换 - 执行兑换: 用户 77347599 商品 2023062912210 兑换失败，可以自行确认。
07-05 19:00:00 [DEBUG] mys_goods_tool.api | 网络请求返回: {"data":null,"message":"","retcode":1028}
07-05 19:00:00 [ERROR] mys_goods_tool.exchange_mode | 用户 77347599 - 【原神】原石*60 - 线程 1 - 兑换失败
07-05 19:00:00 [INFO] mys_goods_tool.api | 米游币商品兑换 - 执行兑换: 用户 77347599 商品 2023062912210 兑换失败，可以自行确认。
07-05 19:00:00 [DEBUG] mys_goods_tool.api | 网络请求返回: {"data":null,"message":"","retcode":1028}
07-05 19:00:00 [ERROR] mys_goods_tool.exchange_mode | 用户 77347599 - 【原神】原石*60 - 线程 2 - 兑换失败
07-05 19:00:00 [INFO] mys_goods_tool.api | 米游币商品兑换 - 执行兑换: 用户 77347599 商品 2023062912693 兑换失败，可以自行确认。
07-05 19:00:00 [DEBUG] mys_goods_tool.api | 网络请求返回: {"data":null,"message":"","retcode":1028}
07-05 19:00:00 [ERROR] mys_goods_tool.exchange_mode | 用户 77347599 - 【原神】原石*150 - 线程 2 - 兑换失败
全部是1028错误，同时我在七点整点进兑换，结果遇到了一个滑动验证码，等验证通过自然东西就没了。
系统环境是Win11，网络延迟大概在40ms左右，之前在学校10ms，不知道会不会有这方面的影响

**Comment by Ljzd-PRO** [comment]
与 #129 情况相同，目前还没修复

**Comment by hhhggg233** [comment]
> 与 #129 情况相同，目前还没修复

好的，坐等大佬修复。辛苦了

**Comment by Ljzd-PRO** [comment]
又尝试修复了一下，可以试一下：https://github.com/Ljzd-PRO/Mys_Goods_Tool/actions/runs/5479310149

解决了几个可能的问题：
- `device_config`, `salt_config` 如果程序默认值更新了但是用户数据文件用的还是旧的，旧的配置还是会覆盖掉默认值的问题
- 兑换请求的Headers里增加`Referer`，修改`Origin`, `Content-Type`
- Headers的`User-Agent`里的App版本号部分和`x-rpc-app_version`不一致的问题

除了Headers里没有`x-rpc-verify_key`, `x-rpc-device_fp`，其他应该都和米游社App里实际发出的请求一致了。

**Comment by fanlu** [comment]
@Ljzd-PRO  https://github.com/Ljzd-PRO/Mys_Goods_Tool/blob/d77e46369f32b21ac8e2690fb5cb41f2e81db7cc/mys_goods_tool/user_data.py#L345 这个salt是不是要随着`X_RPC_APP_VERSION`的变化而变化？ 从 https://github.com/UIGF-org/mihoyo-api-collect/issues/1 看，好像每个版本都有一个独立的salt

**Comment by Ljzd-PRO** [comment]
> @Ljzd-PRO
> 
> https://github.com/Ljzd-PRO/Mys_Goods_Tool/blob/d77e46369f32b21ac8e2690fb5cb41f2e81db7cc/mys_goods_tool/user_data.py#L345
> 
> 这个salt是不是要随着`X_RPC_APP_VERSION`的变化而变化？ 从 [UIGF-org/mihoyo-api-collect#1](https://github.com/UIGF-org/mihoyo-api-collect/issues/1) 看，好像每个版本都有一个独立的salt

兑换请求不需要DS，因此和salt无关。

**Comment by Ljzd-PRO** [comment]
应该是和device_fp、verify_key有关，看一些别的请求 device_fp有一定的影响

**Comment by hhhggg233** [comment]
使用2.1.0pre测试兑换返回的是2102，商品库存不足，可我记得之前不是返回没到兑换时间吗？

**Comment by Ljzd-PRO** [comment]
> 使用2.1.0pre测试兑换返回的是2102，商品库存不足，可我记得之前不是返回没到兑换时间吗？

应该都是有可能的。

**Comment by fanlu** [comment]
<img width="839" alt="image" src="https://github.com/Ljzd-PRO/Mys_Goods_Tool/assets/1620875/ae0e9841-025a-43a2-b528-f18d9eabae2b">
今天用2.1.0还是1028

**Comment by hhhggg233** [comment]
+1,还是1028

**Comment by Ljzd-PRO** [comment]
我这里显示的是不在兑换时间，60原石，-2106

**Comment by Ljzd-PRO** [comment]
是用的原来的配置文件还是自动创建新的重新登录

**Comment by hhhggg233** [comment]
> 是用的原来的配置文件还是自动创建新的重新登录

用的原来的，多了device_fp

**Comment by hhhggg233** [comment]
> 我这里显示的是不在兑换时间，60原石，-2106

兑换到了吗？

**Comment by Ljzd-PRO** [comment]
> > 我这里显示的是不在兑换时间，60原石，-2106
> 
> 兑换到了吗？

没有

**Comment by Remsk0202** [comment]
现在我手动在米哈游抢原石会弹出来一个滑块身份验证，是因为被检测使用脚本的原因吗？

**Comment by Ljzd-PRO** [comment]
> 现在我手动在米哈游抢原石会弹出来一个滑块身份验证，是因为被检测使用脚本的原因吗？

估计是了，看来加了人机验证😂

**Comment by Ljzd-PRO** [comment]
> 现在我手动在米哈游抢原石会弹出来一个滑块身份验证，是因为被检测使用脚本的原因吗？

你有抓包吗

**Comment by Remsk0202** [comment]
> > 现在我手动在米哈游抢原石会弹出来一个滑块身份验证，是因为被检测使用脚本的原因吗？
> 
> 你有抓包吗

我今天抢原石是脚本+手动抢双线进行，脚本日志报的还是1028，然后今天手动点进去兑换的时候给我弹出来人机验证了，错失60原石

**Comment by Ljzd-PRO** [comment]
App里有提示让你选择游戏账号吗？先完成选择再遇到人机验证吗

**Comment by Remsk0202** [comment]
> App里有提示让你选择游戏账号吗？先完成选择再遇到人机验证吗

APP里并没有让我选择游戏账号，只是我7:00点进去兑换的时候弹出来人机验证了，给我一下子干蒙了，以前我手动抢原石没有遇见过这种情况的

**Comment by hhhggg233** [comment]
> App里有提示让你选择游戏账号吗？先完成选择再遇到人机验证吗

我上次也是遇到了验证 那次是1.4.5和2.0.5双开了 这次就开了2.1.0 手动抢到的

**Comment by Remsk0202** [comment]
> > App里有提示让你选择游戏账号吗？先完成选择再遇到人机验证吗
> 
> 我上次也是遇到了验证 那次是1.4.5和2.0.5双开了 这次就开了2.1.0 手动抢到的

你今天手动抢到了原石吗？我的脚本用的2.0.4版本的

**Comment by hhhggg233** [comment]
> > > App里有提示让你选择游戏账号吗？先完成选择再遇到人机验证吗
> > 
> > 
> > 我上次也是遇到了验证 那次是1.4.5和2.0.5双开了 这次就开了2.1.0 手动抢到的
> 
> 你今天手动抢到了原石吗？我的脚本用的2.0.4版本的

是的 我蹲路由旁边抢的 2.0.5及以前的都没必要用了 请求里面少东西

**Comment by Remsk0202** [comment]
> > > > App里有提示让你选择游戏账号吗？先完成选择再遇到人机验证吗
> > > 
> > > 
> > > 我上次也是遇到了验证 那次是1.4.5和2.0.5双开了 这次就开了2.1.0 手动抢到的
> > 
> > 
> > 你今天手动抢到了原石吗？我的脚本用的2.0.4版本的
> 
> 是的 我蹲路由旁边抢的 2.0.5及以前的都没必要用了 请求里面少东西

2.0.5版本的脚本我使用不了，可以正常发验证码但是输入验证码后告诉我网络异常登录失败

**Comment by Ljzd-PRO** [comment]
也尝试删除原来的用户数据文件，让程序创建新的来用，虽然新版本应该默认还是会采用默认的Headers，并在添加兑换计划的时候获取device_fp。但不排除哪里有点Bug。

**Comment by Ljzd-PRO** [comment]
> > > > > App里有提示让你选择游戏账号吗？先完成选择再遇到人机验证吗
> > > > 
> > > > 
> > > > 我上次也是遇到了验证 那次是1.4.5和2.0.5双开了 这次就开了2.1.0 手动抢到的
> > > 
> > > 
> > > 你今天手动抢到了原石吗？我的脚本用的2.0.4版本的
> > 
> > 
> > 是的 我蹲路由旁边抢的 2.0.5及以前的都没必要用了 请求里面少东西
> 
> 2.0.5版本的脚本我使用不了，可以正常发验证码但是输入验证码后告诉我网络异常登录失败

v2.1.0-beta.1 有修复这个Bug

**Comment by Ljzd-PRO** [comment]
短时间应该是找不到什么好的解决方案了，可能最多就做个网页端兑换模式，到了兑换时间自动显示人机验证界面，然后拼手速完成验证。这个issue我转到讨论区里了

**Reply by hhhggg233** [reply]
基本出人机就寄了 别人秒兑换咱还得拉个条 直接无了😂

**Reply by kuronekoda** [reply]
我也是，上上个星期我就发现了，忘了说这事，开着软件自动抢+手枪，软件反1028，手抢弹了个滑条把我干沉默了，愣了几秒才滑错失了，会不会是咱们经常用脚本的被米忽悠打标签了

**Reply by Ljzd-PRO** [reply]
也可能所有人都要完成这个人机验证

**Reply by pbxs666** [reply]
近几周自动抢+手动抢。碰到过2次弹人机验证，每次都没点到，因为它有个很离谱的设定，人机验证窗口没有防误点措施，只要点到验证窗口之外，就会关闭这个窗口，我也是很懵逼，而且验证窗口的判定区域和右下角的“确认兑换”按钮没有任何的重叠区域。抢的时候那手速😂，弹出验证窗口基本就反应不过来，还在疯狂对着“确认兑换”按钮输出，验证窗口闪一下就过去了😅。
![图片1](https://github.com/Ljzd-PRO/Mys_Goods_Tool/assets/47630168/fa6e0636-5962-4d1d-8b2c-a84b5b2b98df)

去年抢的时候还没有这个东西。
最近米开始在这方面较上劲了。就单说那个实时便笺就不停的触发风控，生怕第三方拿走它的数据🙄。

**Reply by Coolerds** [reply]
并不是每个人都需要人机验证，今天兑换实物成功了。最新版可以正常使用，就是在获取验证码时有些小bug。

**Reply by dmhfd** [reply]
崩3的？会不会是崩3本身就没有验证码

**Comment by Ljzd-PRO** [comment]
有用iOS或者方便HTTP抓包的，手动兑换前可以开启抓包，结束以后检查一下实际的兑换过程

**Comment by Remsk0202** [comment]
有人抢到原石了吗，我手动又抢失败了，很烦。明明点进去还有1000多份

**Reply by Ljzd-PRO** [reply]
我这边App刚开始是提示请求频繁，之后就是库存不足。脚本那边返回不在兑换时间，代码是-2106

**Reply by Ljzd-PRO** [reply]
也抓包了，似乎是没有其他什么额外的网络请求，没看到人机验证相关的。可能是如果遇到了人机验证，那是兑换请求返回了人机验证的任务，需要完成验证以后才会再发送请求

**Reply by Remsk0202** [reply]
可是脚本都抢不到原石了吗？那究竟是什么样的人才能抢得到原石呢……好烦啊，连续两个版本我都没有抢到原石了

**Reply by kuronekoda** [reply]
抢实体周边我抢了几回都有人机验证，到现在都抢不到，不知道别人抢实体要不要人机验证，如果咱们一直有人机验证那米游币不废了么

**Reply by Ljzd-PRO** [reply]
脚本还可以再改进一下，就是会不断发请求，持续几秒，可能兑换时间后0-1秒这么短的时间它是不认的

**Reply by Ljzd-PRO** [reply]
原石的我没遇到人机验证
